mne/preprocessing/ica.py

```diff
@@ -144,16 +144,24 @@ class ICA(ContainsMixin):
         will be used. Defaults to ``None``; the actual number used when
         executing the :meth:`ICA.fit` method will be stored in the attribute
         ``n_components_`` (note the trailing underscore).
-    max_pca_components : int | None
-        Number of principal components (from the pre-whitening PCA step) that
-        are retained for later use (i.e., for signal reconstruction in
-        :meth:`ICA.apply`; see the ``n_pca_components`` parameter). Use this
-        parameter to reduce the dimensionality of the input data via PCA before
-        any further processing is performed. If ``None``, no  dimensionality
-        reduction occurs and ``max_pca_components`` will equal the number of
-        channels in the :class:`mne.io.Raw`, :class:`mne.Epochs`, or
-        :class:`mne.Evoked` object passed to :meth:`ICA.fit`. Defaults to
-        ``None``.
+    max_pca_components : int | float | None
+        The number of principal components (from the pre-whitening PCA step)
+        that are retained for later use (i.e., for signal reconstruction in
+        `~mne.preprocessing.ICA.apply`; see the ``n_pca_components``
+        parameter). Use this parameter to reduce the dimensionality of the
+        input data via PCA before any further processing is performed.
+        If :class:`float` between 0 and 1, the number of components with
+        cumulative explained variance less than ``max_pca_components`` will be
+        used. If ``1.0`` or ``None``, no dimensionality reduction occurs and
+        ``max_pca_components`` will equal the number of channels in the
+        `~mne.io.Raw` or `~mne.Epochs` object passed to
+        `~mne.preprocessing.ICA.fit`. Defaults to ``None``. When calling
+        `~mne.preprocessing.ICA.fit`, the attribute ``max_pca_components_``
+        will be set to the absolute number of retained PCA components.
+
+        .. versionchanged:: 0.22
+            Added support for float to select components by cumulative
+            explained variance.
```

Comment:
Wouldn't you typically want to just use `max_pca_components=None` and `n_pca_components=float`? Or would this behave differently?

Reply:
I hadn't even thought about this one. I need to try and see how it works on our data.

If that **does** work, I'm wondering why we'd need `max_pca_components` **at all**?!

Reply:
Because we could simply by default fit PCA with number of components == number of channels, and then use `n_pca_components` and `n_components` to select only a subset of them. I just looked at the code and it seems as if `max_pca_components` could be dropped entirely… but I'd need to do some more testing.

Reply:
> Wouldn't you typically want to just use `max_pca_components=None` and `n_pca_components=float`? Or would this behave differently?

I just tested. Would behave differently:

- `max_pca_components=None` --> feed **all** PCs to ICA
- `n_pca_components=X` --> use `X` PCs for signal reconstruction after ICA run

**these settings are orthogonal**

Let's flip this:

- `max_pca_components=X` --> only keep `X` PCs around, and feed them to ICA
- `n_pca_components=None` --> use **all** retained PCs for signal reconstruction after ICA run (i.e., `X` PCs in our case)


-------

This still begs the question, IMHO, why the h* we need `max_pca_components`. What one could to is:

- drop `max_pca_components`
- select all PCs to keep via `n_pca_components`
- select a subset of those PCs to feed into ICA via `n_components`

Reply:
This just to say: I think this PR is good to merge as-is, and we should maybe discuss the ICA API in a new issue.

Reply:
(This conversation **again** shows that the current API is too complicated)